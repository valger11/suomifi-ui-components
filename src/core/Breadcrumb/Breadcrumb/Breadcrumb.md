```jsx
import { Breadcrumb, BreadcrumbLink } from 'suomifi-ui-components';

<Breadcrumb aria-label="breadcrumb">
  <BreadcrumbLink href="/">Frontpage</BreadcrumbLink>
  <BreadcrumbLink href="/sub">Sub page</BreadcrumbLink>
  <BreadcrumbLink current>Sub sub page</BreadcrumbLink>
</Breadcrumb>;
```
