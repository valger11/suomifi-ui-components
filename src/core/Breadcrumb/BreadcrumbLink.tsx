import React, { Fragment } from 'react';
import { Link, LinkProps } from '../Link/Link';
import { Icon } from '../Icon/Icon';
import { HtmlSpan } from '../../reset';
import classnames from 'classnames';

const baseClassName = 'fi-breadcrumb';
const linkClassName = `${baseClassName}_link`;
const iconClassName = `${baseClassName}_icon`;

export interface BreadcrumbLinkProps extends Omit<LinkProps, 'href'> {
  /** Indicating the link is the current page */
  current?: boolean;
  /** url for the link */
  href?: string;
}

export const BreadcrumbLink = ({
  current,
  children,
  className,
  href = '',
  ...passProps
}: BreadcrumbLinkProps) => {
  if (!!current) {
    return (
      <HtmlSpan className={className} aria-current="location">
        {children}
      </HtmlSpan>
    );
  }

  return (
    <Fragment>
      <Link
        {...passProps}
        className={classnames(linkClassName, className)}
        href={href}
      >
        {children}
      </Link>
      <Icon icon="linkBreadcrumb" className={iconClassName} />
    </Fragment>
  );
};
