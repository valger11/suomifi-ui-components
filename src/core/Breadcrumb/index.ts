export { Breadcrumb, BreadcrumbProps } from './Breadcrumb/Breadcrumb';
export {
  BreadcrumbLink,
  BreadcrumbLinkProps,
} from './BreadcrumbLink/BreadcrumbLink';
