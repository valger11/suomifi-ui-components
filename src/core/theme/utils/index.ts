export { margin, padding, spacingModifiers } from './spacing';
