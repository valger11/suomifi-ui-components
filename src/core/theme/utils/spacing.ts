import {
  SuomifiTheme,
  SpacingProp,
  SpacingDesignTokens,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

type SpacingDesignTokensKeys = keyof SpacingDesignTokens;

export const spacingTokensKeys = Object.keys(
  defaultSuomifiTheme.spacing,
) as SpacingDesignTokensKeys[];

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

/**
 * Spacing type
 * @typedef {String} spacingType
 * @param {String} type 'padding' or 'margin'
 */
/**
 * Spacing tokens
 * @typedef {String} spacingTokens
 * @param {String} t token for top, or top and bottom, or all
 * @param {String} r token for right, or right and left
 * @param {String} b token for bottom
 * @param {String} l token for left
 */
/**
 * Set margin or padding pased on tokens
 * @param {Object} theme suomifiTheme
 * @return {(spacingType) => (spacingTokens) => String}
 */
const space = (theme: SuomifiTheme) => (type: 'padding' | 'margin') => (
  t?: SpacingProp,
  r?: SpacingProp,
  b?: SpacingProp,
  l?: SpacingProp,
) =>
  !!t
    ? `${!!t ? `${type}-top: ${spaceVal(theme)(t)};` : ''}
    ${!!r ? `${type}-right: ${spaceVal(theme)(r)};` : ''}
    ${!!b ? `${type}-bottom: ${spaceVal(theme)(b)};` : ''}
    ${!!l ? `${type}-left:${spaceVal(theme)(l)};` : ''}`
    : '';

/**
 * Set margin based on theme
 * @param {Object} theme
 */
export const margin = (theme: SuomifiTheme) => space(theme)('margin');
/**
 * Set padding based on theme
 * @param {Object} theme
 */
export const padding = (theme: SuomifiTheme) => space(theme)('padding');

/**
 * Create spacing styles for CSS-selector (-xxs, -xs, -s...)
 * TODO: this should be in suomifiTheme? or integrated somehow with that
 * @param {Object} tokens Design tokens
 * @return {(spacingType) => (selector: String) => String}
 */
export const spacingModifiers = (theme: SuomifiTheme) => (
  spacing: 'padding' | 'margin' | 'margin-bottom',
) => (selector: string) =>
  spacingTokensKeys.reduce(
    (ret, k) =>
      `${ret}${selector}-${k}{
        ${spacing}: ${spaceVal(theme)(k)}
  } `,
    '',
  );
