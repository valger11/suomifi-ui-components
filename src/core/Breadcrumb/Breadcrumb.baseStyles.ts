import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { cssValueToString } from '../../utils/css';
import { nav, list, listItem, font } from '../theme/reset';

export const baseStyles = (theme: SuomifiTheme) => css`
  ${nav(theme)}
  ${font(theme)('bodyTextSmall')}
  background-color: ${theme.colors.whiteBase};
  height: 1.5em;

  & .fi-breadcrumb {
    &_list {
      ${list(theme)}
      ${font(theme)('bodyTextSmall')}
      margin: 0;
      padding: 0;
    }
    &_item {
      ${listItem(theme)}
      ${font(theme)('bodyTextSmall')}
      float: left;
      color: ${theme.colors.depthDark1};
    }
    &_item,
    &_link,
    &_icon {
      font-size: ${cssValueToString(
        theme.values.typography.bodyTextSmall.fontSize,
      )};
    }
    &_icon {
      transform: translateY(0.2em);
      margin: 0 ${theme.spacing.insetXs};
      & .fi-icon-base-fill {
        fill: ${theme.colors.depthDark1};
      }
    }
  }
`;
