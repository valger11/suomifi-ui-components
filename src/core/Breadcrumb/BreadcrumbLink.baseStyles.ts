import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { font } from '../theme/reset';
import { cssValueToString } from '../../utils/css';

export const baseStyles = (theme: SuomifiTheme) => css`
  &.fi-breadcrumb-link {
    font-size: ${cssValueToString(
      theme.values.typography.bodyTextSmall.fontSize,
    )};
    display: inline-block;
    & .fi-breadcrumb-link_link {
      ${font(theme)('bodyTextSmall')}
      &--current {
        color: ${theme.colors.depthDark1};
      }
    }

    & .fi-breadcrumb-link_icon {
      transform: translateY(0.2em);
      margin: 0 ${theme.spacing.insetXs};
      fill: ${theme.colors.depthDark1};
    }
  }
`;
