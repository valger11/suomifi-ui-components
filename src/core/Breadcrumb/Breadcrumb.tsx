import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { getConditionalAriaProp } from '../../utils/aria';
import { BaseLinkProps } from '../Link/BaseLink/BaseLink';
import { BreadcrumbLink, BreadcrumbLinkProps } from './BreadcrumbLink';
import { HtmlLi, HtmlNav, HtmlNavProps, HtmlOl } from '../../reset';
import { baseStyles } from './Breadcrumb.baseStyles';
import { SuomifiTheme, SuomifiThemeConsumer } from '../theme';

type BreadcrumbVariant = 'default' | 'link';
const baseClassName = 'fi-breadcrumb';
const listClassName = `${baseClassName}_list`;
const itemClassName = `${baseClassName}_item`;

export interface BreadcrumbProps extends HtmlNavProps {
  /** Name for the breadcrumb like 'Breadcrumb' */
  'aria-label': string;
  /** Custom classname to extend or customize */
  className?: string;
  /**
   * Link element displayed content
   */
  children?: ReactNode;
  /**
   * 'default' | 'link'
   * @default default
   */
  variant?: BreadcrumbVariant;
}

type VariantBreadcrumbProps =
  | BreadcrumbProps
  | (BaseLinkProps & { variant?: BreadcrumbVariant });

const breadcrumbItems = (children: ReactNode) =>
  React.Children.map(children, (child) => (
    <HtmlLi className={itemClassName}>{child}</HtmlLi>
  ));

class BaseBreadcrumb extends Component<
  BreadcrumbProps & { theme: SuomifiTheme }
> {
  render() {
    const { className, theme, children, ...passProps } = this.props;
    return (
      <HtmlNav {...passProps} className={classnames(baseClassName, className)}>
        <HtmlOl className={listClassName}>{breadcrumbItems(children)}</HtmlOl>
      </HtmlNav>
    );
  }
}

const StyledBreadcrumb = styled(BaseBreadcrumb)`
  ${({ theme }) => baseStyles(theme)};
`;

/**
 * <i class="semantics" />
 * Used for navigation path
 */
export class Breadcrumb extends Component<VariantBreadcrumbProps> {
  render() {
    const { variant, 'aria-label': ariaLabel, ...passProps } = this.props;
    if (variant === 'link') {
      return <BreadcrumbLink {...(passProps as BreadcrumbLinkProps)} />;
    }
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <StyledBreadcrumb
            theme={suomifiTheme}
            {...passProps}
            {...getConditionalAriaProp('aria-label', [ariaLabel])}
          />
        )}
      </SuomifiThemeConsumer>
    );
  }
}
