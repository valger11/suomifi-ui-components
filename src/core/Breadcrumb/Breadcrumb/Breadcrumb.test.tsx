import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';
import { Breadcrumb } from './Breadcrumb';
import { BreadcrumbLink } from '../BreadcrumbLink/BreadcrumbLink';

const TestBreadcrumb = (
  <Breadcrumb data-testid="Breadcrumb" aria-label="breadcrumb">
    <BreadcrumbLink href="/">Frontpage</BreadcrumbLink>
    <BreadcrumbLink href="/sub">Sub page</BreadcrumbLink>
    <BreadcrumbLink current={true}>Sub sub page</BreadcrumbLink>
  </Breadcrumb>
);

test('calling render with the same component on the same container does not remount', () => {
  const BreadcrumbRendered = render(TestBreadcrumb);
  const { container } = BreadcrumbRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestBreadcrumb));
