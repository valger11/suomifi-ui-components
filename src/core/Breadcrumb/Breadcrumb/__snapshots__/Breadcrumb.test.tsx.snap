// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c6:hover,
.c6:active,
.c6:focus,
.c6:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,16%);
  display: block;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  height: 1.5em;
}

.c1 .fi-breadcrumb_list {
  color: hsl(0,0%,16%);
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 .fi-breadcrumb_item {
  color: hsl(0,0%,16%);
  list-style: none;
  float: left;
}

.c7 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:hover,
.c7:active,
.c7:focus,
.c7:focus-within {
  color: hsl(212,63%,45%);
}

.c7:focus,
.c7:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:focus {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c7:hover,
.c7:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7:visited {
  color: hsl(284,36%,45%);
}

.c8 {
  display: inline-block;
  vertical-align: baseline;
}

.c8.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c8.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c5.fi-breadcrumb-link {
  font-size: 16px;
  display: inline-block;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--current {
  color: hsl(202,7%,40%);
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_icon {
  -webkit-transform: translateY(0.2em);
  -ms-transform: translateY(0.2em);
  transform: translateY(0.2em);
  margin: 0 4px;
}

.c1 .fi-breadcrumb_icon .fi-icon-base-fill {
  fill: hsl(202,7%,40%);
}

<nav
  aria-label="breadcrumb"
  class="c0 fi-breadcrumb c1"
  data-testid="Breadcrumb"
>
  <ol
    class="c2 fi-breadcrumb_list"
  >
    <li
      class="c3 fi-breadcrumb_item"
    >
      <a
        class="c4 fi-link c5 fi-breadcrumb_link"
        href="/"
      >
        Frontpage
      </a>
      <svg
        aria-hidden="true"
        class="fi-icon c6 fi-breadcrumb_icon"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="fi-icon-base-fill"
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
          fill="#292929"
          fill-rule="evenodd"
        />
      </svg>
    </li>
    <li
      class="c3 fi-breadcrumb_item"
    >
      <a
        class="c4 fi-link c5 fi-breadcrumb_link"
        href="/sub"
      >
        Sub page
      </a>
      <svg
        aria-hidden="true"
        class="fi-icon c6 fi-breadcrumb_icon"
        focusable="false"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="fi-icon-base-fill"
          d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
          fill="#292929"
          fill-rule="evenodd"
        />
      </svg>
    </li>
    <li
      class="c3 fi-breadcrumb_item"
    >
      <span
        class="c4 c5 fi-breadcrumb-link"
      >
        <span
          aria-current="location"
          class="c4 fi-breadcrumb-link_link fi-breadcrumb-link_link--current"
        >
          Sub sub page
        </span>
      </span>
    </li>
  </ol>
</nav>
`;
